src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1291,18 +1328,63 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
+            // to ensure there are no data races at play.
+            //
+            //   1. No active async buffer op and numBytes fits in existing buffer:
+            //       Simply perform a memory copy operation and return synchronously.
+            //
+            //   2. Active async buffer op and numBytes <= _bufferSize
+            //       Since there's an active buffer operation, a new buffer needs to be allocated
+            //       to write to in order to ensure the existing buffer isn't modified mid-write.
+            //       The incoming write can then be copied into the new buffer.
+            //
+            //   3. No active async buffer and numBytes too large for buffer
+            //       If there's buffered data, issue a flush operation and then directly issue the
+            //       incoming write since it can't be buffered.
+            //
+            //   4. Active async buffer op and numBytes too large for buffer
+            //       If there's buffered data, attach it to the existing async operation using Task.WhenAll
+            //       and then directly issue the incoming write.
+            //
+            //       Note: since _writePos is reset synchronously when calling FlushWriteAsync, it can be
+            //       assumed that if _writePos > 0 and there's an active async flush operation, the buffer
+            //       was reallocated since that operation was initiated.
+
+            // Case 1 - no active buffer op & fits in remaining size
+            int remainingBuffer = _bufferSize - _writePos;
+            if(!ActiveBufferOperation() && numBytes <= remainingBuffer)
             {
-                if (_writePos == 0) _buffer = new byte[_bufferSize];
+                if (_buffer == null) _buffer = new byte[_bufferSize];
+
                 Buffer.BlockCopy(array, offset, _buffer, _writePos, numBytes);
                 _writePos += numBytes;
-                
                 return Task.CompletedTask;
             }
 
+            // Case 2 - active buffer op & fits in new buffer
+            else if (ActiveBufferOperation() && numBytes <= _bufferSize)
+            {
+                _buffer = new byte[_bufferSize];
+
+                Buffer.BlockCopy(array, offset, _buffer, 0, numBytes);
+                _writePos = numBytes;
+                return Task.CompletedTask;
+            }
+
+            // Case 3 & 4 - The incoming write can't be handled with a synchronous
+            // memcpy, so issue a write to disk, flushing if necessary
+
             if (_writePos > 0)
-                FlushWrite(false);
+            {
+                Task flushTask = FlushWriteAsync(cancellationToken);
+                if (ActiveBufferOperation())
+                    _activeBufferOperation = Task.WhenAll(_activeBufferOperation, flushTask);
+                else
+                    _activeBufferOperation = flushTask;
+            }
```

Comment:
If there was buffered data that we flush, I think we need to return a Task.WhenAll that includes the flushTask and the WriteInternalCoreAsync task.  That way, if either the write or the flush (which is logically part of the write given how this is structured) fail, the caller will know that there was some sort of problem and the stream may not represent what they hoped it would.